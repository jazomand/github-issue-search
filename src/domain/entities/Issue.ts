import Label from './Label';

class Issue{
    constructor (
        public id: number,
        public title: string,
        public labels: Label[],
        public url: string,
        public repositoryUrl: string,
        public state: string,
        public createdAt: Date,
        public updatedAt: Date|null,
        public closedAt: Date|null,
        public body: string
    ) {}
};

export default Issue;