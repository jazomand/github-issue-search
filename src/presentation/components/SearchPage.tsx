import React, {useState} from 'react';
import { AppBar, Button, Tab, Tabs, Typography, Box } from '@material-ui/core';
import Paper from '@material-ui/core/Paper';
import SearchWithoutRedux from './SearchWithoutRedux';
import SearchWithRedux from './SearchWithRedux';

export default function SearchPage(props: any) {
    const [value, setValue] = React.useState(0);

    const handleChange = (event: any, newValue: any) => {
        setValue(newValue);
    };

    const TabPanel = ({ children, value, index} : {children: any, value: number, index: number}) => {
        return (
          <div
            role="tabpanel"
            hidden={value !== index}
            id={`tabpanel-${index}`}
            aria-labelledby={`tab-${index}`}
          >
            {value === index && (
              <Box p={3}>
                {children}
              </Box>
            )}
          </div>
        );
    };

    return (
        <div >
            <Paper style={{margin: '5vh', padding: '3vh'}} elevation={3} >
                <Typography variant="h4" style={{marginBottom: '1vh'}}>Github facebook/react repository issues</Typography>
                <Typography variant="subtitle2">Press: "ESC" key to unselect</Typography>
                <Typography variant="subtitle2">Press: "UP & DOWN" keys to select an issue.</Typography>
                <Typography variant="subtitle2">Press: "ENTER" key to select a word from autocomplete.</Typography>
                <AppBar position="static" style={{marginTop: '2vh'}}>
                    <Tabs value={value} onChange={handleChange} aria-label="tabs">
                        <Tab label="Search without Redux" />
                        <Tab label="Search with Redux" />
                    </Tabs>
                </AppBar>
                <TabPanel value={value} index={0}>
                    <SearchWithoutRedux {...props} />
                </TabPanel>
                <TabPanel value={value} index={1}>
                    <SearchWithRedux {...props} />
                </TabPanel>
            </Paper>
        </div>
    );
}