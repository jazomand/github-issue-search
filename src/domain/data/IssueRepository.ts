import Issue from '../entities/Issue';

export default interface IssueRepository { 
    getAll: ()=>Promise<Array<Issue>>
    search: (text: string)=>Promise<Array<Issue>>
}