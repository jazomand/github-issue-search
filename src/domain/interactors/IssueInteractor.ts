import Issue from '../entities/Issue';
import IssueRepository from '../data/IssueRepository';

export default class IssueInteractor {

    constructor(private issueRepository : IssueRepository) {}

    public getAll = (): Promise<Array<Issue>> => {
        return this.issueRepository.getAll();
    }

    // Not currently used because it search for full strings only
    public search = (text: string): Promise<Array<Issue>> => {
        return this.issueRepository.search(text);
    }

    public getMatchingWords = (text: string, issues: Array<Issue>) : Array<string> => {
        let words = [] as Array<string>;
        if (text.length < 3){
            return words;
        }
        const wordMap = new Map<string, number>();
        const addToWordMap = (word: string) => {
            let count: number | undefined = wordMap.get(word);
            if (!count){
                count = 1;
            }else{
                count++;
            }
            wordMap.set(word, count);
        };
        issues.forEach(issue =>{
            (this.transformToSpaces(issue.body).split(' ') ?? [])
                .filter(w => w.toLowerCase().indexOf(text.toLowerCase()) != -1)
                .forEach(word => addToWordMap(word));
            
            (this.transformToSpaces(issue.title).split(' ') ?? [])
                .filter(w => w.toLowerCase().indexOf(text.toLowerCase()) != -1)
                .forEach(word => addToWordMap(word));
        });
        wordMap.forEach((value, key)=>{
            words.push(key);
        });
        words.sort((a, b) => (wordMap.get(b) ?? 0)-(wordMap.get(a) ?? 0));
        if (words.length > 10){
            words = words.slice(0,10);
        }
        return words;
    }

    public filterIssues = (text: string, issues: Issue[]) => {
        return issues.filter(issue => 
            (issue.body?.indexOf(text) != -1) || issue.title?.indexOf(text) != -1
        );
    };

    private transformToSpaces = (text: string) => {
        var ch1 = String.fromCharCode(10);
        var ch2 = String.fromCharCode(13);
        if (text.replaceAll){
            return text.replaceAll(ch1, ' ').replaceAll(ch2, '');
        }else{
            return text.split(ch1).join(' ').split(ch2).join('');
        }
    }
}