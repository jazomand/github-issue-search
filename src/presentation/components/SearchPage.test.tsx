import React from 'react';
import { 
    act,
  render,
} from '@testing-library/react';
import SearchPage from './SearchPage';
import IssueInteractor from '../../domain/interactors/IssueInteractor';
import IssueRepository from '../../domain/data/IssueRepository';
import Issue from '../../domain/entities/Issue';

class MockIssueRepository implements IssueRepository {
    public getAll = () :Promise<Issue[]> => {
        return Promise.resolve([]);
    }
    public search = (text: string) : Promise<Issue[]> => {
        return Promise.resolve([]);
    }
}

class MockIssueInteractor extends IssueInteractor {
    constructor(){
        super(new MockIssueRepository());
    }
    public getAll = () : Promise<Issue[]> => Promise.resolve([]);
    public getMatchingWords = (text: string, issues: Issue[]) : string[] => [];
}

test('SearchPage renders correctly', async () => {
    const promise = Promise.resolve();
    const mockIssueInteractor : IssueInteractor = new MockIssueInteractor();
    const {asFragment} = render(<SearchPage issueInteractor={mockIssueInteractor} />);
    expect(asFragment()).toMatchSnapshot();
    await act(() => promise);
});