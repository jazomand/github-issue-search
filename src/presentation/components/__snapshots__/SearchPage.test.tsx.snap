// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPage renders correctly 1`] = `
<DocumentFragment>
  <div>
    <div
      class="MuiPaper-root MuiPaper-elevation3 MuiPaper-rounded"
      style="margin: 5vh; padding: 3vh;"
    >
      <h4
        class="MuiTypography-root MuiTypography-h4"
        style="margin-bottom: 1vh;"
      >
        Github facebook/react repository issues
      </h4>
      <h6
        class="MuiTypography-root MuiTypography-subtitle2"
      >
        Press: "ESC" key to unselect
      </h6>
      <h6
        class="MuiTypography-root MuiTypography-subtitle2"
      >
        Press: "UP & DOWN" keys to select an issue.
      </h6>
      <h6
        class="MuiTypography-root MuiTypography-subtitle2"
      >
        Press: "ENTER" key to select a word from autocomplete.
      </h6>
      <header
        class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
        style="margin-top: 2vh;"
      >
        <div
          class="MuiTabs-root"
        >
          <div
            class="MuiTabs-scroller MuiTabs-fixed"
            style="overflow: hidden;"
          >
            <div
              aria-label="tabs"
              class="MuiTabs-flexContainer"
              role="tablist"
            >
              <button
                aria-selected="true"
                class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit Mui-selected"
                role="tab"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiTab-wrapper"
                >
                  Search without Redux
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-selected="false"
                class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
                role="tab"
                tabindex="-1"
                type="button"
              >
                <span
                  class="MuiTab-wrapper"
                >
                  Search with Redux
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
            <span
              class="PrivateTabIndicator-root-1 PrivateTabIndicator-colorSecondary-3 MuiTabs-indicator"
              style="left: 0px; width: 0px;"
            />
          </div>
        </div>
      </header>
      <div
        aria-labelledby="tab-0"
        id="tabpanel-0"
        role="tabpanel"
      >
        <div
          class="MuiBox-root MuiBox-root-5"
        >
          <div>
            <div
              aria-expanded="false"
              class="MuiAutocomplete-root Mui-focused"
              role="combobox"
            >
              <div
                class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-focused Mui-focused"
                  data-shrink="true"
                  for="search-without-redux"
                  id="search-without-redux-label"
                >
                  Search title & body
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiAutocomplete-inputRoot MuiInputBase-fullWidth Mui-focused Mui-focused MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
                >
                  <input
                    aria-autocomplete="list"
                    aria-invalid="false"
                    autocapitalize="none"
                    autocomplete="off"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd MuiInputBase-inputTypeSearch"
                    id="search-without-redux"
                    spellcheck="false"
                    type="search"
                    value=""
                  />
                  <div
                    class="MuiAutocomplete-endAdornment"
                  />
                  <fieldset
                    aria-hidden="true"
                    class="PrivateNotchedOutline-root-6 MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="PrivateNotchedOutline-legendLabelled-8 PrivateNotchedOutline-legendNotched-9"
                    >
                      <span>
                        Search title & body
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
              role="progressbar"
              style="width: 40px; height: 40px;"
            >
              <svg
                class="MuiCircularProgress-svg"
                viewBox="22 22 44 44"
              >
                <circle
                  class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
                  cx="44"
                  cy="44"
                  fill="none"
                  r="20.2"
                  stroke-width="3.6"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
      <div
        aria-labelledby="tab-1"
        hidden=""
        id="tabpanel-1"
        role="tabpanel"
      />
    </div>
  </div>
</DocumentFragment>
`;
